### PR TITLE
Hide unnecessary itemmeta from clients

### LINKED ISSUE #6316: Block hacks' ability to see player's inventory
### Is your feature request related to a problem?

Yes, 2 weeks ago someone reported an user for using a hack (WurstClient) for using an `.invsee` command to see their armor. I know this may not sound like a big deal, but for instance, we have custom enchantments that can be combined in certain ways and require some knowledge about that, people can just _spy_ others and 'steal' their sets.

https://github.com/Wurst-Imperium/Wurst7/blob/7f453e13e44e834b5a3c6a94efeadcd73771cefe/src/main/java/net/wurstclient/commands/InvseeCmd.java#L62

### Describe the solution you'd like.

I actually have no idea how this would be implemented.

### Describe alternatives you've considered.

I was looking for a way to block this with packets but the Open Window packet is not being sent for player inventory.

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

You see your inventory but the armor and first slot from hotbar are of the other player's.

### PR DESCRIPTION
**Pull request by NoahvdAa**
This PR aims to fix #6316 by allowing server owners to disable sending unnecessary itemmeta (that could provide an advantage to cheat clients) to clients. To be specific:
- item count is set to 1 and damage is reset to 0.
- display.Lore and display.Name are removed.
- If Enchantments is set, the list is replaced with a list containing an empty tag. (So the client still renders the item with the enchantment glow).
- AttributeModifiers are removed completely.

This can be toggled with `hide-itemmeta-from-clients` in paper.yml.
Please let me know if I've missed any tags.

### CODE REVIEW

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index 95d8cf449bf0439fddaf319246fef51e1570b68e..2a67e33dbe802518bf7fbe500df8dad550d3bfb8 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -501,6 +501,11 @@ public class PaperConfig {
+         deobfuscateStacktraces = getBoolean("settings.loggers.deobfuscate-stacktraces", deobfuscateStacktraces);
+     }
+ 
++    public static boolean hideItemmetaFromClients = false;
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: This should be a world option

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
++            if(itemstack1.getTag() != null && PaperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                itemstack1.setDamageValue(0);
++                // We can't just strip out display, leather helmets still use the display.color tag.
++                if(nbt.get("display") != null) {
++                    nbt.getCompound("display").remove("Lore");
++                    nbt.getCompound("display").remove("Name");
++                }
++                if(nbt.get("Enchantments") != null) {
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: What if it’s empty?

**Review by Proximyst — CHANGES_REQUESTED** [review]
Certain enchantments like sharpness also need to stay as they give extra effects.

**Comment by NoahvdAa** [comment]
> Certain enchantments like sharpness also need to stay as they give extra effects.

What does the client use these enchantments for? Is it for the `enchanted_hit`particle?

**Comment by Proximyst** [comment]
> > Certain enchantments like sharpness also need to stay as they give extra effects.
> 
> What does the client use these enchantments for? Is it for the `enchanted_hit`particle?

Yeah, I would assume that’s the one. It’s some blue hit effect.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 3]
```diff
@@ -4,48 +4,41 @@ Date: Tue, 3 Aug 2021 17:28:27 +0200
...
 +                itemstack1.setCount(1);
 +                itemstack1.setDamageValue(0);
 +                // We can't just strip out display, leather helmets still use the display.color tag.
-+                if(nbt.get("display") != null) {
++                if(nbt.get("display") instanceof CompoundTag) {
 +                    nbt.getCompound("display").remove("Lore");
 +                    nbt.getCompound("display").remove("Name");
 +                }
-+                if(nbt.get("Enchantments") != null) {
++                if(nbt.get("Enchantments") instanceof ListTag && ((ListTag) nbt.get("Enchantments")).size() != 0) {
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Java 16 is here! You can use pattern matching here.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+         equipment.forEach((enumitemslot, itemstack) -> {
+             ItemStack itemstack1 = itemstack.copy();
+ 
++            // Paper start - hide unnecessary itemmeta from clients.
++            if(itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                itemstack1.setDamageValue(0);
++                // We can't just strip out display, leather helmets still use the display.color tag.
++                if(nbt.get("display") instanceof CompoundTag) {
```
Comment (kennytv) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Don't get the tag 3 times.

**Review by kennytv — CHANGES_REQUESTED** [review]

**Comment by NoahvdAa** [comment]
> > > Certain enchantments like sharpness also need to stay as they give extra effects.
> > 
> > 
> > What does the client use these enchantments for? Is it for the `enchanted_hit`particle?
> 
> Yeah, I would assume that’s the one. It’s some blue hit effect.

I just did a quick test, and after hitting a pig with a sharpness sword, the particles still appear after for both players.

**Comment by Owen1212055** [comment]
That would be correct. Those particles are sent via an entity animation packet.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
++            if(itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                itemstack1.setDamageValue(0);
++                // We can't just strip out display, leather helmets still use the display.color tag.
++                if(nbt.get("display") instanceof CompoundTag displayTag) {
++                    displayTag.remove("Lore");
++                    displayTag.remove("Name");
++                }
++                if(nbt.get("Enchantments") instanceof ListTag enchantmentsTag && enchantmentsTag.size() != 0) {
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]:
```suggestion
+                if(nbt.get("Enchantments") instanceof ListTag enchantmentsTag && !enchantmentsTag.isEmpty()) {
```

**Review by Proximyst — APPROVED** [review]

**Review by MiniDigger — APPROVED** [review]
I too love pattern matching and ruining cheaters day

**Comment by Machine-Maker** [comment]
For ItemStacks, would it be simpler to construct a new tag and only put what was needed into it? A 1 count, and CustomModelData, and an empty enchant of it was enchanted? Instead of removing everything else.

**Comment by MiniDigger** [comment]
I thought about that too, especially since there might be all kind of other data on the nbt

oh, already done as I am writing this, nice :D

**Comment by NoahvdAa** [comment]
> For ItemStacks, would it be simpler to construct a new tag and only put what was needed into it? A 1 count, and CustomModelData, and an empty enchant of it was enchanted? Instead of removing everything else.

This seems like a better approach, I've gone and implemented that.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
++            // Paper start - hide unnecessary itemmeta from clients.
++            if(itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag oldNbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                itemstack1.setDamageValue(0);
++
++                CompoundTag newNbt = new CompoundTag();
++
++                if (oldNbt.contains("CustomModelData"))
++                    newNbt.put("CustomModelData", oldNbt.get("CustomModelData"));
```
Comment (MiniDigger) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: please add brackets for readability here

**Comment by NoahvdAa** [comment]
By the way, would it be a good idea to enable this by default? Most servers probably don't need this to be disabled.

**Review by Machine-Maker — APPROVED** [review]
> By the way, would it be a good idea to enable this by default? Most servers probably don't need this to be disabled.

I think it should be enabled by default, and servers that use some weirdness can disable it.

Also, I bet we are missing something the client needs but I'm sure someone will report that quickly.

**Comment by NoahvdAa** [comment]
> > By the way, would it be a good idea to enable this by default? Most servers probably don't need this to be disabled.
> 
> I think it should be enabled by default, and servers that use some weirdness can disable it.

Done!

> Also, I bet we are missing something the client needs but I'm sure someone will report that quickly.

Looks like I forgot about SkullOwner, should be added now! For reference, I used [this list](https://minecraft.fandom.com/wiki/Tutorials/Command_NBT_tags#Items) to check for missing tags.

**Comment by iGabyTM** [comment]
Hei @kennytv, do you think you can take a look at this? 😅

**Comment by qixils** [comment]
missing CustomPotionColor, Potion (fallback to determine color i presume)

**Comment by Owen1212055** [comment]
I don't think this should be enabled by default, as there are so many items with little visual tags that can change the appearance of the item, lots of which are currently right now missing. I do not think people will be able to know "oh, the reason why my compass doesn't glow for other people is because of "hide-itemmeta-from-clients" being enabled, I should disable that". For the servers that need it, they can enable it. However, for the majority of servers, like creative servers, or those being noncompetitive, it seems quite pointless for this information to be hidden.

Missing (For Reference: Yarn ModelPredicateProviderRegistry and ItemColors)
Compass: LodestoneTracked, LodestoneDimension, LodestonePos
Maps: MapColor
Elytra: Don't hide durability if it's 1, as it has a special cracked icon
Bundle: Changes depending on amount of items inside the bundle
Crossbow: Changes if it has a firework/arrow in it or not.
Light: Changes depending on the blockstate stored (light level) 
Firework Star: Changes depending on what colors are stored in it

**Comment by qixils** [comment]
Tbh durability shouldn't be hidden at all, considering resource packs can utilize it to render damaged items

**Comment by Owen1212055** [comment]
> Tbh durability shouldn't be hidden at all, considering resource packs can utilize it to render damaged items

So I am guessing that for pvp related servers that type of information is very important. "Oh his helmet is about to break"
I think that this should stay, meaning that players that are on "competitive" servers (or servers that don't want to share this information)  can easily disable this type of info via this config. But I do not think that this should be enabled by default, as to prevent this type of data from being accidentally hidden.

**Review by kennytv — CHANGES_REQUESTED** [review]
Blocking while the remaining issues have not been addressed. I also don't think this should be enabled by default, but that's open for further discussion.

**Comment by NoahvdAa** [comment]
I've gone and added all missing NBT tags. I've added tags that don't need a "special treatment" to a list that gets looped over, so we don't have the same code copy pasted ten times for different tags.

---

> I do not think people will be able to know "oh, the reason why my compass doesn't glow for other people is because of "hide-itemmeta-from-clients" being enabled, I should disable that".

Maybe a readme, similiar to `allow-permanent-block-break-exploits-readme` in paper.yml would work?

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 7]
```diff
@@ -21,29 +21,43 @@ index c334f29c69c1e6e3fe55cd6695e7df400cf36058..2f3a66c1b652b059bfb3cf0b421dfe98
...
-index 2b7eeb5659b1083ef550eb9feb0b7ba8a92a92e3..62411159a25247395bc48fc48384b6590349768f 100644
+index 2b7eeb5659b1083ef550eb9feb0b7ba8a92a92e3..48975a3365f66ae90cffe7d8643864aec04938be 100644
 --- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
 +++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
-@@ -3041,6 +3041,41 @@ public abstract class LivingEntity extends Entity {
+@@ -3035,12 +3035,81 @@ public abstract class LivingEntity extends Entity {
+ 
+     }
+ 
++    private final List<String> shownNBTTags = java.util.Collections.unmodifiableList(List.of("CustomModelData", "SkullOwner", "CustomPotionColor", "Potion", "LodestoneTracked", "LodestoneDimension", "LodestonePos", "Charged", "ChargedProjectiles", "Explosion")); // Paper
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Static & `List#of` returns an unmodifiable list

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Comment by NoahvdAa** [comment]
> > I do not think people will be able to know "oh, the reason why my compass doesn't glow for other people is because of "hide-itemmeta-from-clients" being enabled, I should disable that".
> 
> Maybe a readme, similiar to `allow-permanent-block-break-exploits-readme` in paper.yml would work?

Would something like this work?

`hide-itemmeta-from-clients-readme: When this setting is enabled, the server will not send information about items that may provide an unfair advantage to other players, such as enchantments, durability and the content of bundles and shulkers. Items will still appear as normal in your own inventory.`

**Comment by Owen1212055** [comment]
Also missing banners on shields and the normal banner item.

**Comment by NoahvdAa** [comment]
> Also missing banners on shields and the normal banner item.

Good catch, fixed!

**Review by kennytv — CHANGES_REQUESTED** [review]
Sorry to be throwing you around that much, but we didn't realize how much would have to be manually kept when the suggestion was made to reconstruct the nbt instead of deleting what you *don't* want.

To reduce future maintenance burden and produce a simpler diff, please go back to the deleting approach. Also please keep this disabled by default and remove the config readme hack.

**Comment by NoahvdAa** [comment]
Should I make a PaperDocs PR for this too? Most people probably won't know what `hide-itemmeta-from-clients` means.

**Comment by kennytv** [comment]
Yes.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+@@ -3041,6 +3041,49 @@ public abstract class LivingEntity extends Entity {
+         equipment.forEach((enumitemslot, itemstack) -> {
+             ItemStack itemstack1 = itemstack.copy();
+ 
++            // Paper start - hide unnecessary itemmeta from clients.
++            if (itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                // Only show damage values for elytas.
++                itemstack1.setDamageValue(itemstack1.is(Items.ELYTRA) && itemstack1.getDamageValue() >= itemstack1.getMaxDamage() - 1 ? itemstack1.getDamageValue() : 0);
```
Comment (kennytv) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Put that into a proper if statement, looks fine otherwise

**Review by kennytv — APPROVED** [review]

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
++++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+@@ -3041,6 +3041,49 @@ public abstract class LivingEntity extends Entity {
+         equipment.forEach((enumitemslot, itemstack) -> {
+             ItemStack itemstack1 = itemstack.copy();
+ 
++            // Paper start - hide unnecessary itemmeta from clients.
++            if (itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
++                // Only show damage values for elytas.
```
Comment (Camotoy) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: elyta -> elytra

Comment (NoahvdAa) [reply patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Good catch, thanks!

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 10]
```diff
@@ -21,19 +21,23 @@ index c334f29c69c1e6e3fe55cd6695e7df400cf36058..edfd20abcabc9c2046fa419799ce089d
...
  
 +            // Paper start - hide unnecessary itemmeta from clients.
 +            if (itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
 +                CompoundTag nbt = itemstack1.getTag();
 +                itemstack1.setCount(1);
-+                // Only show damage values for elytas.
-+                itemstack1.setDamageValue(itemstack1.is(Items.ELYTRA) && itemstack1.getDamageValue() >= itemstack1.getMaxDamage() - 1 ? itemstack1.getDamageValue() : 0);
++                // Only show damage values for elytras.
++                if (itemstack1.is(Items.ELYTRA) && itemstack1.getDamageValue() >= itemstack1.getMaxDamage() - 1) {
++                    itemstack1.setDamageValue(itemstack1.getDamageValue());
```
Comment (kennytv) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: 🦧

Comment (NoahvdAa) [reply patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: 🦧

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
++++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+@@ -3041,6 +3041,53 @@ public abstract class LivingEntity extends Entity {
+         equipment.forEach((enumitemslot, itemstack) -> {
+             ItemStack itemstack1 = itemstack.copy();
+ 
++            // Paper start - hide unnecessary itemmeta from clients.
++            if (itemstack1.getTag() != null && this.level.paperConfig.hideItemmetaFromClients) {
++                CompoundTag nbt = itemstack1.getTag();
++                itemstack1.setCount(1);
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: We should preserve showing plural for resource packs using OptiFine's stuff. Just doing e.g. `itemstack1.setCount(itemstack1.getCount() > 1 ? 2 : 1)` is enough.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
++                // Only show damage values for elytras.
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Should also preserve damage values for the same reason; I can't imagine it being a big deal for cheaters anyhow.

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
++                    int sizeUsed = 0;
++                    for (org.bukkit.inventory.ItemStack item : bundleMeta.getItems()) {
++                        int scale = 64 / item.getMaxStackSize();
++                        sizeUsed += scale * item.getAmount();
++                    }
++                    // Now we add a single fake item that uses the same amount of slots as all other items.
++                    ListTag items = new ListTag();
++                    CompoundTag fakeItem = new CompoundTag();
++                    fakeItem.putInt("Count", sizeUsed);
++                    fakeItem.putString("id", "minecraft:paper");
```
Comment (Proximyst) [inline patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Hehe, I see what you did here. Either way, you should use the `Items.PAPER` constant for this.

**Review by Proximyst — DISMISSED** [review]

File: patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch (hunk 12 above)
Comment (NoahvdAa) [reply patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: I feel like damage values can give a significant advantage in combat, knowing when your opponent's armor is about to break can drastically change the outcome of a fight.

Comment (qixils) [reply patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: Durability textures are configurable in vanilla resource packs, although I think it's fine to break this functionality given that the config option is disabled by default. This should be noted in the documentation though.

Comment (NoahvdAa) [reply patches/server/0738-Hide-unnecessary-itemmeta-from-clients.patch]: I've added a note to the docs pr for this @ PaperMC/PaperDocs#101

**Comment by iGabyTM** [comment]
Hey guys, any idea if this will be implemented any time soon? 😅

**Comment by iGabyTM** [comment]
Hey guys, any idea if this will be implemented any time soon? 😅

**Comment by Cryptite** [comment]
Suggestion to also consider this change: https://github.com/Cryptite/Slice/commit/b6e39c709659e613b2f1b5483768d5c27c34cc66

**Comment by NoahvdAa** [comment]
Added config changes discussed in #paper-dev

**Review by Machine-Maker — APPROVED** [review]

**Comment by Owen1212055** [comment]
Link to convo? I’m curious why item meta is hidden on default again.

**Comment by Machine-Maker** [comment]
@Owen1212055 the basics of it is, having the setting enabled by default shouldn't have any gameplay/visual effect. If it does, its either a bug with this patch and can be fixed, OR someone is doing something really janky and they can disable it. There's a separate setting for durability that is left on by default because of resource packs.

**Comment by Cryptite** [comment]
It's worth noting this will (pretty sure) break players who use OptiFine CIT resource packs to visually change things based on various item metadata. So that will effectively break that usage for server owners as the default case.

**Comment by iGabyTM** [comment]
> It's worth noting this will (pretty sure) break players who use OptiFine CIT resource packs to visually change things based on various item metadata. So that will effectively break that usage for server owners as the default case.

If this is the case, I agree the change is not really worth =/ fucking hacks 🤣

**Comment by drunderscore** [comment]
@iGabyTM If you don't need it then disable it. My server uses custom CIT textures and I will have to disable it.

If you still want the effects of this PR but, say, want to keep `PublicBukkitValues`, you can edit and build this patch yourself.

**Comment by iGabyTM** [comment]
> @iGabyTM If you don't need it then disable it. My server uses custom CIT textures and I will have to disable it.
> 
> If you still want the effects of this PR but, say, want to keep `PublicBukkitValues`, you can edit and build this patch yourself.

Hmm, what if `PublicBukkitValues` are preserved by default? In my case at least, the problem is the lore not PDC.

**Comment by drunderscore** [comment]
> > @iGabyTM If you don't need it then disable it. My server uses custom CIT textures and I will have to disable it.
> > If you still want the effects of this PR but, say, want to keep `PublicBukkitValues`, you can edit and build this patch yourself.
> 
> Hmm, what if `PublicBukkitValues` are preserved by default? In my case at least, the problem is the lore not PDC.

PDC is usually useless for clients to have. And besides, as MiniDigger and others have said in #6316
> anyways, using display name as a identifier is flawed and wrong. custom model data exists for a reason.

**Comment by drunderscore** [comment]
Actually, I can't think of _any_ use for a client to know about any PDC. Even though I use it myself for Optifine CIT, it's totally wrong and I'll be inclined to properly use custom model data.

**Comment by iGabyTM** [comment]
> Actually, I can't think of _any_ use for a client to know about any PDC. Even though I use it myself for Optifine CIT, it's totally wrong and I'll be inclined to properly use custom model data.

Yeah but some plugins use it or other NBT tags. On my server we use some public plugins and a custom texture pack (Optifine CIT) and to apply the texture on items we use NBT (e.g. AdvancedEnchantments store NBT in books as `ae_book`)

**Comment by drunderscore** [comment]
I do the exact same thing in my plugin and it's totally wrong. Either disable it, patch it out yourself, or _properly use custom model data_.

**Comment by NoahvdAa** [comment]
Rebased

**Comment by NoahvdAa** [comment]
Rebased, should be ready to go

**Review by kennytv — APPROVED** [review]

**Comment by cosmiolotl** [comment]
RIP Chunkbans

**Comment by stijnsmet** [comment]
bye bye chunkbans, hello spigot prisons

**Pull request merged**